sqlserver/check.py

```diff
@@ -165,31 +174,79 @@ def typed_metric(self, dd_name, sql_name, base_name, user_type, sql_type, instan
 
-        return cls(dd_name, sql_name, base_name,
+        return cls(self._get_connector(instance), dd_name, sql_name, base_name,
                    metric_type, instance_name, tag_by, self.log)
 
+    def _get_connector(self, instance):
+        connector = instance.get('connector', self.connector)
+        if connector != self.connector:
+            if not connector.lower() in self.valid_connectors:
+                self.log.warning("Invalid database connector %s using default %s" ,
+                     connector, self.connector)
+                connector = self.connector
+            else:
+                self.log.info("Overriding default connector for %s with %s", instance['host'], connector)
+        return connector
+    
     def _get_access_info(self, instance):
         ''' Convenience method to extract info from instance
         '''
-        host = instance.get('host', '127.0.0.1,1433')
+        dsn = instance.get('dsn')
+        host = instance.get('host')
         username = instance.get('username')
         password = instance.get('password')
-        database = instance.get('database', self.DEFAULT_DATABASE)
-        return host, username, password, database
+        database = instance.get('database')
+        driver = instance.get('driver')
+        if not dsn:
+            if not host:
+                host = '127.0.0.1,1433'
+            if not database:
+                database = self.DEFAULT_DATABASE
+            if not driver:
+                driver = self.DEFAULT_DRIVER
+        return dsn, host, username, password, database, driver
 
     def _conn_key(self, instance):
         ''' Return a key to use for the connection cache
         '''
-        host, username, password, database = self._get_access_info(instance)
-        return '%s:%s:%s:%s' % (host, username, password, database)
+        dsn, host, username, password, database, driver = self._get_access_info(instance)
+        return '%s:%s:%s:%s:%s:%s' % (dsn, host, username, password, database, driver)
+
+    def _conn_string_odbc(self, instance=None, conn_key=None):
+        ''' Return a connection string to use with odbc
+        '''
+        if instance:
+            dsn, host, username, password, database, driver = self._get_access_info(instance)
+        elif conn_key:
+            dsn, host, username, password, database, driver = conn_key.split(":")
+        #conn_str = 'Provider=SQLOLEDB;Data Source=%s;Initial Catalog=%s;' \
+        #    % (host, database)
+        conn_str = ''
+        if dsn:
+            conn_str = 'DSN=%s;' % (dsn)
+
+        if driver:
+            conn_str += 'DRIVER={%s};' % (driver)
+        if host:
+            conn_str += 'Server=%s;' % (host)
+        if database:
+            conn_str += 'Database=%s;' % (database)
 
-    def _conn_string(self, instance=None, conn_key=None):
+        if username:
+            conn_str += 'UID=%s;' % (username)
+        self.log.debug("Connection string (before password) %s" % conn_str)
```

Comment:
Nit: you can use the log string interpolation (it's safer): `self.log.debug("Connection string (before password) %s", conn_str)`